crates/core/src/util/slow.rs

```diff
@@ -0,0 +1,250 @@
+    pub fn with_defaults() -> Self {
+        Self {
+            incremental_updates: None,
+            subscriptions: None,
+            queries: Some(Duration::from_millis(THRESHOLD_QUERIES_MILLIS)),
+        }
+    }
+
+    /// Sets the threshold for incremental updates.
+    pub fn with_incremental_updates(mut self, duration: Duration) -> Self {
+        self.incremental_updates = Some(duration);
+        self
+    }
+    /// Sets the threshold for subscriptions.
+    pub fn with_subscriptions(mut self, duration: Duration) -> Self {
+        self.subscriptions = Some(duration);
+        self
+    }
+    /// Sets the threshold for general queries.
+    pub fn with_queries(mut self, duration: Duration) -> Self {
+        self.queries = Some(duration);
+        self
+    }
+
+    pub fn for_queries(self, sql: &str) -> SlowQuery {
+        SlowQuery::query(self, sql)
+    }
+    pub fn for_subscriptions(self, sql: &str) -> SlowQuery {
+        SlowQuery::query(self, sql)
+    }
+    pub fn for_incremental_updates(self, sql: &str) -> SlowQuery {
+        SlowQuery::incremental_updates(self, sql)
+    }
+}
+
+impl Default for SlowQueryConfig {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Represents `threshold` for [SlowQuery].
+pub enum Threshold {
+    IncrementalUpdates(Option<Duration>),
+    Subscriptions(Option<Duration>),
+    Queries(Option<Duration>),
+}
+
+/// Start the recording of a `sql` with a specific [Threshold].
+pub struct SlowQuery<'a> {
+    /// The SQL statement of the query.
+    sql: &'a str,
+    /// The start time of the query execution.
+    start: Instant,
+    /// Which [Threshold] to use.
+    threshold: Threshold,
+}
+
+impl<'a> SlowQuery<'a> {
+    pub fn new(sql: &'a str, threshold: Threshold) -> Self {
+        Self {
+            sql,
+            start: Instant::now(),
+            threshold,
+        }
+    }
+
+    /// Creates a new [SlowQuery] instance for general queries.
+    pub fn query(config: SlowQueryConfig, sql: &'a str) -> Self {
+        Self::new(sql, Threshold::Queries(config.queries))
+    }
+
+    /// Creates a new [SlowQuery] instance for subscriptions.
+    pub fn subscription(config: SlowQueryConfig, sql: &'a str) -> Self {
+        Self::new(sql, Threshold::Subscriptions(config.subscriptions))
+    }
+
+    /// Creates a new [SlowQuery] instance for incremental updates.
+    pub fn incremental_updates(config: SlowQueryConfig, sql: &'a str) -> Self {
+        Self::new(sql, Threshold::IncrementalUpdates(config.queries))
+    }
+
+    /// Log as `tracing::warn!` the query if it exceeds the threshold.
+    pub fn log(&self) -> Option<Duration> {
+        let (kind, dur) = match self.threshold {
+            Threshold::IncrementalUpdates(dur) => ("IncrementalUpdates", dur),
+            Threshold::Subscriptions(dur) => ("Subscriptions", dur),
+            Threshold::Queries(dur) => ("Queries", dur),
```

Comment:
I would rename the printable `kind` as follows:

"IncrementalUpdates" -> "Incremental update query"
"Subscriptions" -> "Subscription query"
"Queries" -> "Ad hoc query" or "One off query"